Add nnc connections to grid

atgeirr opened the pull request:
With this, non-neighbouring connections (NNCs) are passed to the CpGrid class and incorporated as (fake) faces in the grid structure. This is done in order to allow downstream code to handle NNCs without any code changes.

Internally, the NNC faces have the lowest indices, the order of faces within the grid should now be NNC, I, J, K. This preserves the internal convention that the K faces are the two last faces of any cell. This is only of interest in the internal implementation, as we do not iterate over faces/intersections directly in any case.

Work remains to use the feature from Flow, in particular to ensure the transmissibilities are correctly ordered.

@totto82: I have removed (commented out) the code adding NNCs from PINCH, replacing it with the more generic code below. It believe it will result in having for such cells both a K face (that is an internal boundary face) and an NNC face, instead of just a K face that connects to the other cell. I do not think it should have any ill effects, but it would be good if you could check this. What test case is the most relevant here?

@blattms: I think you could review this PR now for snags and bugs, in any case it is not really testable until companion PRs are added. I will still fire off Jenkins though.

atgeirr commented:
jenkins build this please

atgeirr commented:
jenkins build this please

blattms commented on opm/grid/cpgrid/Intersection.cpp:
```diff
@@ -121,12 +125,17 @@ int Intersection::indexInInside() const
+        // For nnc faces we return the otherwise unused value -1.
+        // The Dune grid interface is essentially meaningless
+        // for non-neighbouring "intersections".
+        return -1;
```
I doubt, that this works with the ewoms discretization.

atgeirr replied:
Indeed, this required a matching change in ecltransmissibilities.hh, as can be seen in the companion PR. It only required that one change though.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -668,9 +680,24 @@ namespace cpgrid
+                if (c1 < 0 || c2 < 0) {
+                    OPM_THROW(std::runtime_error, "NNC connection requested between inactive cells.");
+                }
+                cells[0].setValue(c1, true);
```
Here you silently assume that an NNC between to cells may only occur once.
This will not work in general, because:

1. It is totally legal to specify an NNC between already connected cells.
2. It is totally legal to specify the same NNC multiple times. In that case transmissibilities will be added.
3. A specified NNC might also occur from PINCH.

atgeirr replied:
You are correct, I will amend to make sure it works in those three cases.

blattms replied:
I tried to understand why the test of the companion PR in ewoms fails.

Could it be that we now end up with two intersections in case an NNC was created via PINCH? From the commented code below it seems that there is an intersection between the old cell pairs (before pinching) that is never removed. Instead we now add a new intersection. Hence we should have one intersection with the old (now inactive cell, that hopefully gets zero transmissibility later) and another one with the correct cell.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -682,7 +709,7 @@ namespace cpgrid
                     ++cellcount;
                 }
-
+                /*
```
Even if the current ewoms code would not need `Intersection::indexInSide()`, we are losing information here because NNCs created by pinch are not treated as regular intersections anymore.

atgeirr replied:
I think that is more precise though: the created faces are not normal K-faces. They will still be intersections though.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -714,7 +742,11 @@ namespace cpgrid
                 int output_face = face_to_output_face[face];
-                f2p.appendRow(fn + fp[output_face], fn + fp[output_face+1]);
+                if (output_face == cpgrid::NNCFace) {
+                    f2p.appendRow(fn, fn);
```
This seems highly dangerous! I do not get what you are doing here. Would you mind to explain? Definitely missing at least a comment.

atgeirr replied:
I am adding an empty row, thereby associating no points with the face. I'll add a comment.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -868,14 +900,22 @@ namespace cpgrid
+                    // and do not have any embedded geometry.
+                    face_normals.push_back({-1e100, -1e100, -1e100});
+                    face_centroids.push_back({-1e100, -1e100, -1e100});
+                    face_areas.push_back(-1e100);
```
This will not work with the ewoms code. Why not try to use some reasonable values?

- Normal should use the two two cell centroids as direction and have length 1
- Centroid should be the midpoint between the two cells
- Area should be one to make sure it can be used with the ewoms discretization code.

atgeirr replied:
I wanted to ensure the values were not actually used, and stand by that: it is better to get crazy results when you try something wrong. However for the area I had to modify it (already in this PR) for the reason you state. I think the normal and centroid should remain invalid. Reasonable values will hide errors.

blattms commented:
> Internally, the NNC faces have the lowest indices, the order of faces within the grid should now be NNC, I, J, K. This preserves the internal convention that the K faces are the two last faces of any cell. This is only of interest in the internal implementation, as we do not iterate over faces/intersections directly in any case.
> 

I am pretty sure that the discretization code iterates over intersections.

> Work remains to use the feature from Flow, in particular to ensure the transmissibilities are correctly ordered.

What do you mean by this? Transmissibilities are in an unorderd_map in ewoms. The key is the pair of cell indices  with small cell index first. Therefore we only need to add the transmissibilities from real NNCs which are currently lacking.

blattms commented:
Well, I am 100% sure that area is used in eclfluxmodule.hh and I think it is
highly probable at least some of the others are too when using the two point
flux approximation.

BTW: Looking at eclfluxmodule.hh I would recommend testing without -DNDEBUG as
I have the feeling that there should be some failing asserts with the current
implementation.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -139,12 +139,19 @@ namespace cpgrid
         }
 
+        // Add explicit NNCs to the NNC map.
+        if (nncs.hasNNC()) {
```
if clause seems superfluous as `nncs.nncdata()` will be empty anyway

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -668,9 +680,24 @@ namespace cpgrid
+            for (const auto& nncpair : nnc) {
+                const int c1 = global_to_local[nncpair.first];
+                const int c2 = global_to_local[nncpair.second];
+                if (c1 < 0 || c2 < 0) {
```
I do not think that specifiying NNCs with inactive cells should be fatal. But I could be wrong

atgeirr commented:
A brief update: I have added tests and fixed some bugs, the issues reported in the review should be addressed (except for those noted in earlier comments). I still want to add a test for the PINCH behaviour and refactor a function that became unreadably long before I consider myself done.

atgeirr commented:
Another thing to note: checking that repeated NNCs behave properly is "too easy" for the grid (the repeats should have no effect). The responsibility for handling the transmissibility properly (adding) in that case is on ecltransmissibility.hh, which cannot be unit tested (only integration tested) for the usual type tag reasons. I'll add an integration test for Flow verifying that the transmissibilities are ok, that will be a separate PR.

atgeirr commented:
I now consider this PR ready for final review. The question of how to handle PINCH nncs should be decided though (add nnc or replace regular connection), it is not obvious to me that doing it one way or the other is better (and I do not know what other simulators do).

I'll run this through jenkins both by itself, and together with the ewoms PR.

atgeirr commented:
jenkins build this please

totto82 commented:
I have tested this on a homemade test case and the results change significantly. I will send the case via email.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -868,14 +1007,27 @@ namespace cpgrid
+                    // we have to set the areas to 1 to avoid trouble.
+                    face_normals.push_back({-1e100, -1e100, -1e100});
+                    face_centroids.push_back({-1e100, -1e100, -1e100});
+                    face_areas.push_back(1.0);
```
I have to reiterate: Are you sure that these values will work? 

I have a feeling that they will break code at [ecltransmissibility.hh#L212-312](/OPM/ewoms/blob/2e1a6f34ca7a58a1aa1b17ec2a5cde36c2c1f38a/ebos/ecltransmissibility.hh#L212-L312). I would guess that this breaks PINCH for cases without NNC.

atgeirr commented:
I have now added proper treatment of PINCH connections. Similar to @totto82's original solution, we again replace a K+ face with one that connects the two cells. In addition, the lower cell's K- face is deleted (as it is replaced with the connecting face). I believe that latter is a small improvement relative to the current treatment of these cells.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -139,12 +145,32 @@ namespace cpgrid
+            // end up with the same {low, high} pair.
+            // The code that computes the transmissibilities is responsible
+            // for ensuring repeated NNC transmissibilities are added.
+            auto low = std::min(single_nnc.cell1, single_nnc.cell2);
```
Thinking about why I would prefer to use
```
auto low  = single_nnc.cell1;
auto high = single_nnc.cell2;
if ( low > high ){
  std:swap(low, high)
}
```
makes me wonder how flow should behave if `low == high` holds. The parser just eats it. I guess that is an error in the deck and we should abort?

atgeirr replied:
I think explicit NNCs between a cell and itself should be flagged as an error in the parser. @joakim-hove do you agree?

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -638,39 +670,126 @@ namespace cpgrid
+                    continue;
+                }
+                auto beg = std::lower_bound(face_cells.begin(), face_cells.end(), std::make_pair(c1, -1));
+                auto end = std::upper_bound(face_cells.begin(), face_cells.end(), std::make_pair(c1 + 1, -1));
```
Do we really want to binary searches over all face_cells?

There should not be two many faces for a cell and therefore the second one could be linear:
```
auto end=std::find(begin, face_cells.end(), [](const auto& pair){return pair->first==c1+1;})
```
As `nnc` is sorted the first one could start with the last end.

blattms commented on opm/grid/cpgrid/processEclipseFormat.cpp:
```diff
@@ -638,39 +670,126 @@ namespace cpgrid
+                auto beg = std::lower_bound(face_cells.begin(), face_cells.end(), std::make_pair(c1, -1));
+                auto end = std::upper_bound(face_cells.begin(), face_cells.end(), std::make_pair(c1 + 1, -1));
+                bool found_same_connection = false;
+                for (auto it = beg; it < end; ++it) {
```
And another find algorithm. Maybe we should merge this with the search for end?

atgeirr commented:
jenkins build this please

atgeirr commented:
jenkins build this please

alfbr commented:
Build has broken for this pull request. @blattms maybe you can help restore building?

alfbr commented:
It turns out that our Boost version was too old for the new test here. This PR requires Boost 1.60 or newer. We have now successfully tested the PR. It gives correct results for NNC keyword. @akva2 , we still have a fall-out on the upscaling tests.

bska commented:
> This PR requires Boost 1.60 or newer.

In that case something's got to give.  We're currently looking for Boost 1.44 during [system probe](https://github.com/OPM/opm-grid/blob/0b9f55571e33a6a7b9d28c180a8df83fe9a21a5b/opm-grid-prereqs.cmake#L26-L27), so we've either got to upgrade OPM's minimum Boost requirement or we need to rewrite the implementation of the new test to support the older versions of Boost.Test.

alfbr commented:
Actually, it is not important to us. Thanks to the absolutely fantastic work done on OPM's build system we have not had build issues for years. Even on archaic Red Hat 6, non-privileged user, deploying out-of-tree and linking in the latest Boost 1.69 out-of-tree just worked today. Just as we are now on Dune 2.5.2 and GCC 7.3, Swapping out CMAKE variables to the correct version number is all it takes. It is probably the latest Ubuntu LTS or Debian stable that dictates minimum versions these days. Building with default GCC on Red Hat 7 (which is version 4.8) broke some time ago without anybody noticing.

alfbr commented:
Just checked. Ubuntu 16.04 is on version 1.58, same as we were yesterday. Latest Debian stable, i.e., Jenkins and benchmarking is on version 1.62. The latest Ubuntu LTS is 18.04 which is on 1.65, but I believe 18.04 has issues with the MPI libraries.

blattms commented:
Turns out supporting older boost is pretty easy. See #356

ytelses commented:
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.035    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.958    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.982    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.917    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=222

alfbr commented:
Hm, performance should not be impacted by this PR. None of the tested models use the NNC keyword. The trigger was unintended though, this PR should be considered with https://github.com/OPM/ewoms/pull/466
